editor/store/selectors.js

```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+ * @property {string[]} keywords          Keywords that can be searched to find this item.
+ * @property {boolean}  isDisabled        Whether or not the user should be prevented from inserting
+ *                                        this item.
+ * @property {number}   utility           How useful we think this item is, between 0 and 3.
+ * @property {number}   frecency          Hueristic that combines frequency and recency.
+ */
+export const getInserterItems = createSelector(
+	( state, parentUID = null ) => {
+		/**
+		 * The second argument used to be {boolean|array} allowedBlockTypes but it is no
+		 * longer necessary.
+		 */
+		if ( isBoolean( parentUID ) || isArray( parentUID ) ) {
+			deprecated( 'allowedBlockTypes', {
+				version: '3.2',
+				plugin: 'Gutenberg',
+			} );
+			parentUID = null;
+		}
 
-	const staticItems = getBlockTypes().map( ( blockType ) =>
-		buildInserterItemFromBlockType( state, allowedBlockTypes, blockType )
-	);
+		const calculateUtility = ( category, count, isContextual ) => {
+			if ( isContextual ) {
+				return INSERTER_UTILITY_HIGH;
+			} else if ( count > 0 ) {
+				return INSERTER_UTILITY_MEDIUM;
+			} else if ( category === 'common' ) {
+				return INSERTER_UTILITY_LOW;
+			}
+			return INSERTER_UTILITY_NONE;
+		};
 
-	const dynamicItems = getSharedBlocks( state ).map( ( sharedBlock ) =>
-		buildInserterItemFromSharedBlock( state, allowedBlockTypes, sharedBlock )
-	);
+		const calculateFrecency = ( time, count ) => {
+			if ( ! time ) {
+				return count;
+			}
+
+			const duration = Date.now() - time;
+			switch ( true ) {
+				case duration < SECONDS_PER_HOUR:
+					return count * 4;
+				case duration < SECONDS_PER_DAY:
+					return count * 2;
+				case duration < SECONDS_PER_WEEK:
+					return count / 2;
+				default:
+					return count / 4;
+			}
+		};
 
-	const items = [ ...staticItems, ...dynamicItems ];
-	return compact( items );
-}
+		const shouldIncludeBlockType = ( blockType ) => {
+			if ( ! hasBlockSupport( blockType, 'inserter', true ) ) {
+				return false;
+			}
 
-function fillWithCommonBlocks( inserts ) {
-	// Filter out any inserts that are associated with a block type that isn't registered
-	const items = inserts.filter( ( insert ) => getBlockType( insert.name ) );
+			if ( ! canInsertBlockType( state, blockType.name, parentUID ) ) {
+				return false;
+			}
 
-	// Common blocks that we'll use to pad out our list
-	const commonInserts = getBlockTypes()
-		.filter( ( blockType ) => blockType.category === 'common' )
-		.map( ( blockType ) => ( { name: blockType.name } ) );
+			return true;
+		};
 
-	const areInsertsEqual = ( a, b ) => a.name === b.name && a.ref === b.ref;
-	return unionWith( items, commonInserts, areInsertsEqual );
-}
+		const buildBlockTypeInserterItem = ( blockType ) => {
+			const id = blockType.name;
+
+			let isDisabled = false;
+			if ( blockType.useOnce ) {
+				isDisabled = getBlocks( state ).some( ( block ) => block.name === blockType.name );
```

Comment:
FYI: Lodash's `_.some` has a nice convenience for this pattern in its object predicates:

```js
isDisabled = some( getBlocks( state ), { name: blockType.name } );
```

https://lodash.com/docs/4.17.10#some

Reply:
👌 changed in ed0b4f104